File: src/routes/reports.py
"""Reports & analytics routes: /api/reports"""
from flask import Blueprint, jsonify
from src.db.init_db import get_db
from src.middleware.auth import require_auth, require_permission

reports_bp = Blueprint("reports", __name__, url_prefix="/api/reports")


@reports_bp.get("/dashboard")
@require_auth
@require_permission("VIEW_REPORTS")
def dashboard():
    """Leadership dashboard KPIs."""
    db = get_db()

    total_active_subs = db.execute("""
        SELECT COUNT(*) FROM subscription
        WHERE renewal_date IS NOT NULL AND renewal_date >= DATE('now')
    """).fetchone()[0]

    unique_active_customers = db.execute("""
        SELECT COUNT(DISTINCT cus_id) FROM subscription
        WHERE renewal_date IS NOT NULL AND renewal_date >= DATE('now')
    """).fetchone()[0]

    current_month_revenue = db.execute("""
        SELECT COALESCE(SUM(amount), 0) FROM payment_transaction
        WHERE payment_status = 'Completed'
          AND strftime('%Y-%m', created_at) = strftime('%Y-%m', 'now')
    """).fetchone()[0]

    current_month_failed = db.execute("""
        SELECT COUNT(*) FROM payment_transaction
        WHERE payment_status = 'Failed'
          AND strftime('%Y-%m', created_at) = strftime('%Y-%m', 'now')
    """).fetchone()[0]

    orders_awaiting = db.execute("""
        SELECT COUNT(*) FROM customer_order
        WHERE order_status IN ('Pending', 'Packed')
    """).fetchone()[0]

    regions_served = db.execute("""
        SELECT COUNT(DISTINCT province) FROM address
    """).fetchone()[0]

    db.close()
    return jsonify({
        "total_active_subscriptions": total_active_subs,
        "unique_active_customers": unique_active_customers,
        "current_month_revenue": float(current_month_revenue),
        "current_month_failed_payments": current_month_failed,
        "orders_awaiting_shipment": orders_awaiting,
        "regions_served": regions_served,
    })


@reports_bp.get("/subscribers/active")
@require_auth
@require_permission("VIEW_REPORTS")
def active_subscribers():
    """Active subscriber summary by plan."""
    db = get_db()
    rows = db.execute("""
        SELECT sp.plan_name,
               COUNT(DISTINCT s.subscription_id) AS active_subscriptions,
               COUNT(DISTINCT s.cus_id)          AS unique_customers
        FROM subscription s
        JOIN subscription_plan sp ON s.plan_id = sp.plan_id
        WHERE s.renewal_date IS NOT NULL AND s.renewal_date >= DATE('now')
        GROUP BY sp.plan_name
    """).fetchall()
    db.close()
    return jsonify([dict(r) for r in rows])


@reports_bp.get("/subscribers/growth")
@require_auth
@require_permission("VIEW_REPORTS")
def subscriber_growth():
    """Cumulative subscriber growth month over month."""
    db = get_db()
    rows = db.execute("""
        SELECT strftime('%Y-%m', s.start_date) AS signup_month,
               COUNT(*) AS new_subscriptions
        FROM subscription s
        GROUP BY signup_month
        ORDER BY signup_month
    """).fetchall()
    db.close()

    # Compute cumulative in Python
    cumulative = 0
    result = []
    for r in rows:
        cumulative += r["new_subscriptions"]
        result.append({**dict(r), "cumulative_subscriptions": cumulative})
    return jsonify(result)


@reports_bp.get("/subscribers/churn")
@require_auth
@require_permission("VIEW_REPORTS")
def churn_rate():
    """Monthly churn rate."""
    db = get_db()
    rows = db.execute("""
        WITH monthly_cancellations AS (
            SELECT strftime('%Y-%m', event_date) AS churn_month,
                   COUNT(*) AS cancellations
            FROM subscription_event
            WHERE event_type = 'Cancelled'
            GROUP BY churn_month
        ),
        monthly_active AS (
            SELECT strftime('%Y-%m', co.created_at) AS active_month,
                   COUNT(DISTINCT s.subscription_id) AS active_subs
            FROM customer_order co
            JOIN subscription s ON co.subscription_id = s.subscription_id
            GROUP BY active_month
        )
        SELECT ma.active_month,
               ma.active_subs,
               COALESCE(mc.cancellations, 0) AS cancellations,
               ROUND(CAST(COALESCE(mc.cancellations,0) AS FLOAT)
                   / ma.active_subs * 100, 2) AS churn_rate_pct
        FROM monthly_active ma
        LEFT JOIN monthly_cancellations mc ON ma.active_month = mc.churn_month
        ORDER BY ma.active_month
    """).fetchall()
    db.close()
    return jsonify([dict(r) for r in rows])


@reports_bp.get("/revenue/by-plan")
@require_auth
@require_permission("VIEW_REPORTS")
def revenue_by_plan():
    """Revenue broken down by subscription plan."""
    db = get_db()
    rows = db.execute("""
        SELECT sp.plan_name,
               COUNT(pt.transaction_id)                          AS total_transactions,
               ROUND(SUM(pt.amount), 2)                         AS gross_revenue,
               ROUND(SUM(pt.tax_amount), 2)                     AS total_tax,
               ROUND(SUM(COALESCE(pt.discount_amount, 0)), 2)   AS total_discounts,
               ROUND(SUM(pt.amount) - SUM(COALESCE(pt.discount_amount,0)), 2) AS net_revenue
        FROM payment_transaction pt
        JOIN subscription s ON pt.subscription_id = s.subscription_id
        JOIN subscription_plan sp ON s.plan_id = sp.plan_id
        WHERE pt.payment_status = 'Completed'
        GROUP BY sp.plan_name
    """).fetchall()
    db.close()
    return jsonify([dict(r) for r in rows])


@reports_bp.get("/revenue/monthly")
@require_auth
@require_permission("VIEW_REPORTS")
def monthly_revenue():
    """Monthly revenue summary."""
    db = get_db()
    rows = db.execute("""
        SELECT strftime('%Y-%m', created_at) AS revenue_month,
               ROUND(SUM(CASE WHEN payment_status='Completed' THEN amount ELSE 0 END), 2) AS completed_revenue,
               ROUND(SUM(CASE WHEN payment_status='Refunded'  THEN amount ELSE 0 END), 2) AS refunded_amount,
               ROUND(SUM(CASE WHEN payment_status='Failed'    THEN amount ELSE 0 END), 2) AS failed_amount,
               ROUND(SUM(CASE WHEN payment_status='Completed' THEN tax_amount ELSE 0 END), 2) AS tax_collected,
               COUNT(CASE WHEN payment_status='Failed' THEN 1 END) AS failed_count
        FROM payment_transaction
        GROUP BY revenue_month
        ORDER BY revenue_month
    """).fetchall()
    db.close()
    return jsonify([dict(r) for r in rows])


@reports_bp.get("/revenue/by-region")
@require_auth
@require_permission("VIEW_REPORTS")
def revenue_by_region():
    """Revenue by geographic region."""
    db = get_db()
    rows = db.execute("""
        SELECT a.province AS region,
               pt.currency_code,
               COUNT(pt.transaction_id) AS total_transactions,
               ROUND(SUM(pt.amount), 2) AS gross_revenue,
               ROUND(SUM(pt.tax_amount), 2) AS total_tax_collected
        FROM payment_transaction pt
        JOIN customer_order co ON pt.order_id = co.order_id
        JOIN subscription s ON co.subscription_id = s.subscription_id
        JOIN customer cu ON s.cus_id = cu.customer_id
        JOIN address a ON cu.customer_id = a.acc_id
        WHERE pt.payment_status = 'Completed'
        GROUP BY a.province, pt.currency_code
        ORDER BY gross_revenue DESC
    """).fetchall()
    db.close()
    return jsonify([dict(r) for r in rows])


@reports_bp.get("/fulfillment/by-release")
@require_auth
@require_permission("VIEW_REPORTS")
def fulfillment_by_release():
    """Units shipped per box type per release period."""
    db = get_db()
    rows = db.execute("""
        SELECT t.theme_name,
               strftime('%Y-%m', br.release_month) AS release_period,
               COUNT(co.order_id) AS total_orders,
               SUM(CASE WHEN co.order_status='Delivered' THEN 1 ELSE 0 END) AS delivered,
               SUM(CASE WHEN co.order_status='Shipped'   THEN 1 ELSE 0 END) AS shipped,
               SUM(CASE WHEN co.order_status='Packed'    THEN 1 ELSE 0 END) AS packed,
               SUM(CASE WHEN co.order_status='Pending'   THEN 1 ELSE 0 END) AS pending,
               SUM(CASE WHEN co.order_status='Failed'    THEN 1 ELSE 0 END) AS failed
        FROM customer_order co
        JOIN box_release br ON co.release_id = br.release_id
        JOIN theme t ON br.theme_id = t.theme_id
        GROUP BY t.theme_name, release_period
        ORDER BY release_period DESC, t.theme_name
    """).fetchall()
    db.close()
    return jsonify([dict(r) for r in rows])


@reports_bp.get("/fulfillment/shipment-performance")
@require_auth
@require_permission("VIEW_REPORTS")
def shipment_performance():
    """Average delivery days per theme per release."""
    db = get_db()
    rows = db.execute("""
        SELECT t.theme_name,
               strftime('%Y-%m', br.release_month) AS release_period,
               COUNT(sh.shipment_id) AS total_shipments,
               SUM(CASE WHEN sh.shipment_status='Delivered' THEN 1 ELSE 0 END) AS delivered,
               ROUND(AVG(CASE
                   WHEN sh.delivered_date IS NOT NULL AND sh.shipped_date IS NOT NULL
                   THEN julianday(sh.delivered_date) - julianday(sh.shipped_date)
               END), 1) AS avg_delivery_days
        FROM shipment sh
        JOIN customer_order co ON sh.order_id = co.order_id
        JOIN box_release br ON co.release_id = br.release_id
        JOIN theme t ON br.theme_id = t.theme_id
        GROUP BY t.theme_name, release_period
        ORDER BY release_period DESC
    """).fetchall()
    db.close()
    return jsonify([dict(r) for r in rows])


@reports_bp.get("/inventory/overview")
@require_auth
@require_permission("VIEW_INVENTORY")
def inventory_overview():
    """Full inventory overview with stock status."""
    db = get_db()
    rows = db.execute("""
        SELECT p.product_name, p.product_price, p.product_cost, w.warehouse_name,
               i.quantity_availability, i.quantity_reserved, i.quantity_damaged,
               (i.quantity_availability - i.quantity_reserved) AS net_available,
               CASE
                   WHEN (i.quantity_availability - i.quantity_reserved) <= 0 THEN 'Out of Stock'
                   WHEN (i.quantity_availability - i.quantity_reserved) < 500 THEN 'Low Stock'
                   ELSE 'In Stock'
               END AS stock_status
        FROM inventory i
        JOIN product p ON i.product_id = p.product_id
        JOIN warehouse w ON i.warehouse_id = w.warehouse_id
        ORDER BY stock_status, p.product_name
    """).fetchall()
    db.close()
    return jsonify([dict(r) for r in rows])


@reports_bp.get("/inventory/damaged")
@require_auth
@require_permission("VIEW_INVENTORY")
def damaged_inventory():
    """Damaged inventory with damage rate percentage."""
    db = get_db()
    rows = db.execute("""
        SELECT w.warehouse_name, p.product_name, i.quantity_damaged,
               ROUND(CAST(i.quantity_damaged AS FLOAT)
                   / (i.quantity_availability + i.quantity_reserved + i.quantity_damaged) * 100, 2)
                   AS damage_rate_pct
        FROM inventory i
        JOIN product p ON i.product_id = p.product_id
        JOIN warehouse w ON i.warehouse_id = w.warehouse_id
        WHERE i.quantity_damaged > 0
        ORDER BY damage_rate_pct DESC
    """).fetchall()
    db.close()
    return jsonify([dict(r) for r in rows])


@reports_bp.get("/themes/popularity")
@require_auth
@require_permission("VIEW_REPORTS")
def theme_popularity():
    """Fandom / theme popularity by customers and orders."""
    db = get_db()
    rows = db.execute("""
        SELECT t.theme_name,
               COUNT(DISTINCT ct.customer_id) AS customers_interested,
               COUNT(DISTINCT co.order_id)    AS total_orders_placed
        FROM theme t
        LEFT JOIN customer_theme ct ON t.theme_id = ct.theme_id
        LEFT JOIN box_release br ON t.theme_id = br.theme_id
        LEFT JOIN customer_order co ON br.release_id = co.release_id
        GROUP BY t.theme_name
        ORDER BY customers_interested DESC
    """).fetchall()
    db.close()
    return jsonify([dict(r) for r in rows])


@reports_bp.get("/audit-log")
@require_auth
@require_permission("VIEW_AUDIT_LOG")
def audit_log():
    """Recent audit log entries."""
    db = get_db()
    limit = min(200, int(request.args.get("limit", 50)) if hasattr(jsonify, '__module__') else 50)
    from flask import request as req
    limit = min(200, req.args.get("limit", 50, type=int))
    rows = db.execute("""
        SELECT al.*, su.username FROM audit_log al
        LEFT JOIN system_user su ON al.user_id = su.user_id
        ORDER BY al.audit_id DESC
        LIMIT ?
    """, (limit,)).fetchall()
    db.close()
    return jsonify([dict(r) for r in rows])

File: functions/src/routes/reports.ts
import { Router, Response } from "express";
import { db } from "../index";
import { requireAuth, requirePermission, AuthRequest } from "../middleware/auth";

export const reportsRouter = Router();

// ── helpers ──────────────────────────────────────────────────────────────────
const monthOf = (iso: string) => iso?.slice(0, 7) ?? "unknown";

/** GET /api/reports/dashboard */
reportsRouter.get(
  "/dashboard",
  requireAuth,
  requirePermission("VIEW_REPORTS"),
  async (_req: AuthRequest, res: Response): Promise<void> => {
    try {
      const today = new Date().toISOString().split("T")[0];
      const thisMonth = today.slice(0, 7);

      const [activeSubs, pendingOrders, txSnap, failSnap, regionsSnap] = await Promise.all([
        db.collection("subscriptions")
          .where("renewal_date", ">=", today).where("auto_renew", "==", true).get(),
        db.collection("orders")
          .where("order_status", "in", ["Pending", "Packed"]).get(),
        db.collection("payment_transactions")
          .where("payment_status", "==", "Completed")
          .where("created_at", ">=", `${thisMonth}-01`).get(),
        db.collection("payment_transactions")
          .where("payment_status", "==", "Failed")
          .where("created_at", ">=", `${thisMonth}-01`).get(),
        db.collection("customers").get(),
      ]);

      const currentRevenue = txSnap.docs.reduce((s, d) => s + (d.data().amount ?? 0), 0);
      const uniqueCustomers = new Set(activeSubs.docs.map((d) => d.data().customer_id)).size;

      // regions = distinct provinces from customer addresses
      const provinces = new Set<string>();
      regionsSnap.docs.forEach((d) => {
        const p = d.data().province;
        if (p) provinces.add(p);
      });

      res.json({
        total_active_subscriptions:  activeSubs.size,
        unique_active_customers:     uniqueCustomers,
        current_month_revenue:       Math.round(currentRevenue * 100) / 100,
        current_month_failed_payments: failSnap.size,
        orders_awaiting_shipment:    pendingOrders.size,
        regions_served:              provinces.size,
      });
    } catch (err) {
      res.status(500).json({ error: String(err) });
    }
  }
);

/** GET /api/reports/subscribers/active */
reportsRouter.get(
  "/subscribers/active",
  requireAuth,
  requirePermission("VIEW_REPORTS"),
  async (_req: AuthRequest, res: Response): Promise<void> => {
    try {
      const today = new Date().toISOString().split("T")[0];
      const snap  = await db.collection("subscriptions")
        .where("renewal_date", ">=", today).where("auto_renew", "==", true).get();

      // Group by plan_id
      const planMap: Record<string, { subs: number; customers: Set<string> }> = {};
      const planIds = new Set<string>();
      snap.docs.forEach((d) => {
        const { plan_id, customer_id } = d.data();
        planIds.add(plan_id);
        if (!planMap[plan_id]) planMap[plan_id] = { subs: 0, customers: new Set() };
        planMap[plan_id].subs++;
        planMap[plan_id].customers.add(customer_id);
      });

      // Fetch plan names
      const planDocs = await Promise.all([...planIds].map((id) =>
        db.collection("subscription_plans").doc(id).get()
      ));
      const planNames: Record<string, string> = {};
      planDocs.forEach((d) => { if (d.exists) planNames[d.id] = d.data()!.plan_name; });

      res.json(Object.entries(planMap).map(([plan_id, v]) => ({
        plan_name:            planNames[plan_id] ?? plan_id,
        active_subscriptions: v.subs,
        unique_customers:     v.customers.size,
      })));
    } catch (err) {
      res.status(500).json({ error: String(err) });
    }
  }
);

/** GET /api/reports/subscribers/growth */
reportsRouter.get(
  "/subscribers/growth",
  requireAuth,
  requirePermission("VIEW_REPORTS"),
  async (_req: AuthRequest, res: Response): Promise<void> => {
    try {
      const snap = await db.collection("subscriptions").orderBy("start_date").get();
      const monthly: Record<string, number> = {};
      snap.docs.forEach((d) => {
        const m = monthOf(d.data().start_date);
        monthly[m] = (monthly[m] ?? 0) + 1;
      });

      let cumulative = 0;
      res.json(Object.entries(monthly).sort().map(([month, count]) => {
        cumulative += count;
        return { signup_month: month, new_subscriptions: count, cumulative_subscriptions: cumulative };
      }));
    } catch (err) {
      res.status(500).json({ error: String(err) });
    }
  }
);

/** GET /api/reports/subscribers/churn */
reportsRouter.get(
  "/subscribers/churn",
  requireAuth,
  requirePermission("VIEW_REPORTS"),
  async (_req: AuthRequest, res: Response): Promise<void> => {
    try {
      // Cancellation events (subcollections)
      const subSnap   = await db.collection("subscriptions").get();
      const cancelMap: Record<string, number> = {};
      const activeMap: Record<string, Set<string>> = {};

      // Active by month from orders
      const orderSnap = await db.collection("orders").get();
      orderSnap.docs.forEach((d) => {
        const m = monthOf(d.data().created_at);
        if (!activeMap[m]) activeMap[m] = new Set();
        activeMap[m].add(d.data().subscription_id);
      });

      // Get cancellations from events subcollections
      await Promise.all(subSnap.docs.map(async (subDoc) => {
        const events = await subDoc.ref.collection("events")
          .where("event_type", "==", "Cancelled").get();
        events.docs.forEach((e) => {
          const m = monthOf(e.data().event_date);
          cancelMap[m] = (cancelMap[m] ?? 0) + 1;
        });
      }));

      res.json(
        Object.entries(activeMap)
          .sort(([a], [b]) => a.localeCompare(b))
          .map(([month, subs]) => ({
            active_month:  month,
            active_subs:   subs.size,
            cancellations: cancelMap[month] ?? 0,
            churn_rate_pct: subs.size > 0
              ? Math.round(((cancelMap[month] ?? 0) / subs.size) * 10000) / 100
              : 0,
          }))
      );
    } catch (err) {
      res.status(500).json({ error: String(err) });
    }
  }
);

/** GET /api/reports/revenue/by-plan */
reportsRouter.get(
  "/revenue/by-plan",
  requireAuth,
  requirePermission("VIEW_REPORTS"),
  async (_req: AuthRequest, res: Response): Promise<void> => {
    try {
      const snap = await db.collection("payment_transactions")
        .where("payment_status", "==", "Completed").get();

      // Need subscription→plan mapping
      const subIds = [...new Set(snap.docs.map((d) => d.data().subscription_id))];
      const subDocs = await Promise.all(subIds.map((id) =>
        db.collection("subscriptions").doc(id).get()
      ));
      const subPlan: Record<string, string> = {};
      subDocs.forEach((d) => { if (d.exists) subPlan[d.id] = d.data()!.plan_id; });

      const planIds = [...new Set(Object.values(subPlan))];
      const planDocs = await Promise.all(planIds.map((id) =>
        db.collection("subscription_plans").doc(id).get()
      ));
      const planName: Record<string, string> = {};
      planDocs.forEach((d) => { if (d.exists) planName[d.id] = d.data()!.plan_name; });

      const grouped: Record<string, { txns: number; gross: number; tax: number; discount: number }> = {};
      snap.docs.forEach((d) => {
        const data = d.data();
        const planId = subPlan[data.subscription_id] ?? "unknown";
        const name   = planName[planId] ?? planId;
        if (!grouped[name]) grouped[name] = { txns: 0, gross: 0, tax: 0, discount: 0 };
        grouped[name].txns++;
        grouped[name].gross    += data.amount    ?? 0;
        grouped[name].tax      += data.tax_amount ?? 0;
        grouped[name].discount += data.discount_amount ?? 0;
      });

      res.json(Object.entries(grouped).map(([plan_name, v]) => ({
        plan_name,
        total_transactions: v.txns,
        gross_revenue: Math.round(v.gross    * 100) / 100,
        total_tax:     Math.round(v.tax      * 100) / 100,
        total_discounts: Math.round(v.discount * 100) / 100,
        net_revenue: Math.round((v.gross - v.discount) * 100) / 100,
      })));
    } catch (err) {
      res.status(500).json({ error: String(err) });
    }
  }
);

/** GET /api/reports/revenue/monthly */
reportsRouter.get(
  "/revenue/monthly",
  requireAuth,
  requirePermission("VIEW_REPORTS"),
  async (_req: AuthRequest, res: Response): Promise<void> => {
    try {
      const snap = await db.collection("payment_transactions").get();
      const monthly: Record<string, {
        completed: number; refunded: number; failed: number; tax: number; failed_count: number;
      }> = {};

      snap.docs.forEach((d) => {
        const { created_at, payment_status, amount, tax_amount } = d.data();
        const m = monthOf(created_at);
        if (!monthly[m]) monthly[m] = { completed:0, refunded:0, failed:0, tax:0, failed_count:0 };
        if (payment_status === "Completed") { monthly[m].completed += amount ?? 0; monthly[m].tax += tax_amount ?? 0; }
        if (payment_status === "Refunded")  { monthly[m].refunded  += amount ?? 0; }
        if (payment_status === "Failed")    { monthly[m].failed    += amount ?? 0; monthly[m].failed_count++; }
      });

      res.json(Object.entries(monthly).sort().map(([month, v]) => ({
        revenue_month:     month,
        completed_revenue: Math.round(v.completed * 100) / 100,
        refunded_amount:   Math.round(v.refunded  * 100) / 100,
        failed_amount:     Math.round(v.failed    * 100) / 100,
        tax_collected:     Math.round(v.tax       * 100) / 100,
        failed_count:      v.failed_count,
      })));
    } catch (err) {
      res.status(500).json({ error: String(err) });
    }
  }
);

/** GET /api/reports/themes/popularity */
reportsRouter.get(
  "/themes/popularity",
  requireAuth,
  requirePermission("VIEW_REPORTS"),
  async (_req: AuthRequest, res: Response): Promise<void> => {
    try {
      const [themes, ctSnap, orderSnap, releaseSnap] = await Promise.all([
        db.collection("themes").get(),
        db.collection("customer_themes").get(),
        db.collection("orders").get(),
        db.collection("box_releases").get(),
      ]);

      const releaseTheme: Record<string, string> = {};
      releaseSnap.docs.forEach((d) => { releaseTheme[d.id] = d.data().theme_id; });

      const ordersByTheme: Record<string, number> = {};
      orderSnap.docs.forEach((d) => {
        const themeId = releaseTheme[d.data().release_id];
        if (themeId) ordersByTheme[themeId] = (ordersByTheme[themeId] ?? 0) + 1;
      });

      const custByTheme: Record<string, number> = {};
      ctSnap.docs.forEach((d) => {
        const { theme_id } = d.data();
        custByTheme[theme_id] = (custByTheme[theme_id] ?? 0) + 1;
      });

      res.json(themes.docs.map((d) => ({
        theme_name:          d.data().theme_name,
        customers_interested: custByTheme[d.id]   ?? 0,
        total_orders_placed:  ordersByTheme[d.id] ?? 0,
      })).sort((a, b) => b.customers_interested - a.customers_interested));
    } catch (err) {
      res.status(500).json({ error: String(err) });
    }
  }
);

/** GET /api/reports/inventory/overview */
reportsRouter.get(
  "/inventory/overview",
  requireAuth,
  requirePermission("VIEW_INVENTORY"),
  async (_req: AuthRequest, res: Response): Promise<void> => {
    try {
      const [invSnap, prodSnap, whSnap] = await Promise.all([
        db.collection("inventory").get(),
        db.collection("products").get(),
        db.collection("warehouses").get(),
      ]);
      const products:   Record<string, Record<string, unknown>> = {};
      const warehouses: Record<string, string> = {};
      prodSnap.docs.forEach((d) => { products[d.id]   = d.data(); });
      whSnap.docs.forEach((d)   => { warehouses[d.id] = d.data().warehouse_name; });

      res.json(invSnap.docs.map((d) => {
        const data = d.data();
        const net  = (data.quantity_availability ?? 0) - (data.quantity_reserved ?? 0);
        return {
          id: d.id,
          product_name:          products[data.product_id]?.product_name,
          product_price:         products[data.product_id]?.product_price,
          warehouse_name:        warehouses[data.warehouse_id],
          quantity_availability: data.quantity_availability,
          quantity_reserved:     data.quantity_reserved,
          quantity_damaged:      data.quantity_damaged,
          net_available: net,
          stock_status:  net <= 0 ? "Out of Stock" : net < 500 ? "Low Stock" : "In Stock",
        };
      }));
    } catch (err) {
      res.status(500).json({ error: String(err) });
    }
  }
);

/** GET /api/reports/audit-log */
reportsRouter.get(
  "/audit-log",
  requireAuth,
  requirePermission("VIEW_AUDIT_LOG"),
  async (req: AuthRequest, res: Response): Promise<void> => {
    try {
      const limit = Math.min(200, parseInt(req.query.limit as string) || 50);
      const snap  = await db.collection("audit_logs")
        .orderBy("created_at", "desc").limit(limit).get();
      res.json(snap.docs.map((d) => ({ id: d.id, ...d.data() })));
    } catch (err) {
      res.status(500).json({ error: String(err) });
    }
  }
);
